sdks/python/apache_beam/io/gcp/dicomio.py

```diff
@@ -0,0 +1,419 @@
+    match = p | fileio.MatchFiles(path)
+    fileio_obj = match | fileio.ReadAll()
+    results = fileio_obj | DicomStoreInstance(input_dict, 'fileio')
+
+  with Pipeline() as p:
+    input_dict = {'project_id': 'abc123', 'type': 'instances',...}
+    f = open("abc.dcm", "rb")
+    dcm_file = f.read()
+    byte_file = p | 'create byte file' >> beam.Create([dcm_file])
+    results = byte_file | DicomStoreInstance(input_dict, 'bytes')
+
+The first example uses a PCollection of fileio objects as input.
+DicomStoreInstance will read DICOM files from the objects and send them
+to a DICOM storage.
+The second example uses a PCollection of byte[] as input. DicomStoreInstance
+will directly send those DICOM files to a DICOM storage.
+Users can also get the operation results in the output PCollection if they want
+to handle the failed store requests.
+"""
+
+# pytype: skip-file
+from __future__ import absolute_import
+
+import apache_beam as beam
+from apache_beam.io.gcp.dicomclient import DicomApiHttpClient
+from apache_beam.transforms import PTransform
+
+
+class DicomSearch(PTransform):
+  """A PTransform used for retrieving DICOM instance metadata from Google
+    Cloud DICOM store. It takes a Pcollection of dicts as input and return
+    a Pcollection of dict as results:
+    INPUT:
+    The input dict represents DICOM web path parameters, which has the following
+    string keys and values:
+    {
+      'project_id': str,
+      'region': str,
+      'dataset_id': str,
+      'dicom_store_id': str,
+      'search_type': str,
+      'params': dict(str,str) (Optional),
+    }
+    Key-value pairs:
+      project_id: Id of the project in which DICOM store locates. (Required)
+      region: Region where the DICOM store resides. (Required)
+      dataset_id: Id of the dataset where DICOM store belongs to. (Required)
+      dicom_store_id: Id of the dicom store. (Required)
+      search_type: Which type of search it is, could only be one of the three
+        values: 'instances', 'series', or 'studies'. (Required)
+      params: A dict of str:str pairs used to refine QIDO search. (Optional)
+        Supported tags in three categories:
+          1. Studies:
+            StudyInstanceUID
+            PatientName
+            PatientID
+            AccessionNumber
+            ReferringPhysicianName
+            StudyDate
+          2. Series: all study level search terms and
+            SeriesInstanceUID
+            Modality
+          3. Instances: all study/series level search terms and
+            SOPInstanceUID
+        e.g. {"StudyInstanceUID":"1","SeriesInstanceUID":"2"}
+    OUTPUT:
+    The output dict wraps results as well as error messages:
+    {
+      'result': a list of dicts in JSON style.
+      'success': boolean value telling whether the operation is successful.
+      'input': detail ids and dicomweb path for this retrieval.
+      'status': status code from the server, used as error message.
+    }
```

Comment:
Yes it does. The main reason that input parameters are saved in the result dictionary is for traceability. Since the healthcare API users don't want to lose any data (maybe some information or x-rays for a patient) when using the pipeline, I put the input parameters in the result dictionary so that they can trace the details of the failed operations.
(Please tell me if there is a better way to do this cause I am not familiar with Beam error handling)